snuba/utils/streams/synchronized.py

```diff
@@ -0,0 +1,203 @@
+from dataclasses import dataclass
+from threading import Event
+from typing import Callable, Mapping, MutableMapping, Optional, Sequence, Set
+
+from snuba.utils.concurrent import Synchronized, execute
+from snuba.utils.streams.consumer import Consumer, ConsumerError, EndOfPartition
+from snuba.utils.streams.types import Message, Partition, Topic, TPayload
+
+
+@dataclass(frozen=True)
+class Commit:
+    __slots__ = ["group", "partition", "offset"]
+
+    group: str
+    partition: Partition
+    offset: int
+
+
+class SynchronizedConsumer(Consumer[TPayload]):
+    def __init__(
+        self,
+        consumer: Consumer[TPayload],
+        commit_log_consumer: Consumer[Commit],
+        commit_log_topic: Topic,
+        commit_log_groups: Set[str],
+    ) -> None:
+        self.__consumer = consumer
+
+        self.__commit_log_consumer = commit_log_consumer
+        self.__commit_log_topic = commit_log_topic
+        self.__commit_log_groups = commit_log_groups
+
+        self.__remote_offsets: Synchronized[
+            Mapping[str, MutableMapping[Partition, int]]
+        ] = Synchronized({group: {} for group in commit_log_groups})
```

Comment:
You need to explicitly type the whole expression because mypy is not able to infer `MutableMapping[Partition, int]` is that correct ?

Reply:
Yeah — if removed, I get this:

```
ted@veneno % python -m mypy --strict snuba/utils/streams/synchronized.py
snuba/utils/streams/synchronized.py:33: error: Need type annotation for '__remote_offsets'
```

I'm not 100% sure it's because of the empty dictionary, but that seems like a plausible explanation.